### PR TITLE
Migrate `src/model/doc` to TypeScript

### PR DESCRIPTION
**Pull request by ChrisAntaki**

### CODE REVIEW

File: src/model/doc.js [hunk 1]
```diff
@@ -1,151 +0,0 @@
...
-   * @return {!Promise}
-   */
-  whenReady() {}
-
-  /**
-   * Adds the element to the fixed layer.
-   * @param {!Element} unusedElement
-   * @return {!Promise}
-   *
-   * This is a no-op for except in AMP on iOS < 13.0.
```
Comment (ChrisAntaki) [inline src/model/doc.js:71]: Note: Only necessary for AMP

File: src/model/doc.js [hunk 2]
```diff
@@ -1,151 +0,0 @@
...
-   */
-  addToFixedLayer(unusedElement) {}
```
Comment (ChrisAntaki) [inline src/model/doc.js:73]: Note: This was only necessary for AMP

**Review by fikus — APPROVED** [review]

File: src/model/doc.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    return this.doc_;
+  }
+
+  getRootElement() {
+    return this.doc_.documentElement;
+  }
+
+  getHead() {
+    // `document.head` always has a chance to be parsed, at least partially.
+    return /** @type {!Element} */ this.doc_.head;
```
[the remark this answers is not in the document]
Comment (ChrisAntaki) [reply src/model/doc.ts]: Nice catch

File: src/model/doc.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isDocumentReady, whenDocumentReady} from '../utils/document-ready';
+
+/**
+ * @interface
```
[the remark this answers is not in the document]
Comment (ChrisAntaki) [reply src/model/doc.ts]: Nice catch

File: src/model/doc.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  isReady(): boolean;
+
+  /**
+   * Resolved when document has been fully constructed.
+   */
+  whenReady(): Promise<void>;
+}
+
+export class GlobalDoc implements Doc {
+  private readonly doc_: Document;
```
[the remark this answers is not in the document]
Comment (ChrisAntaki) [reply src/model/doc.ts:59]: For now, the `_` suffixes are improving minification

**Pull request merged**